### ISSUE TITLE
Can't compile `sf::String` with Libc++ 18 due to undefined `std::char_traits<std::uint8_t>`

### ISSUE THREAD

**Issue by JohelEGP** [SFML/SFML#2670]
## Can't compile `sf::String` with Libc++ 18 due to undefined `std::char_traits<std::uint8_t>`

The C++ standard library doesn't define `std::char_traits<std::uint8_t>`.
It used to compile fine with <https://github.com/llvm/llvm-project/tree/c0abd3814564a568dfc607c216e6407eaa314f46>.

## Your environment

* OS: Arch Linux.
* SFML version: `master`.
* Clang 18 (<https://github.com/llvm/llvm-project/tree/985e399647d591d6130ba6fe08c5b5f6cb87d9f6>).
* Special compiler flags used: `-stdlib=libc++`.

## Steps to reproduce

```bash
cmake -S SFML -B build -DCMAKE_CXX_COMPILER=clang++ -DCMAKE_CXX_FLAGS=-stdlib=libc++
cmake --build build/
```

## Expected behavior

To build just fine.

## Actual behavior

This error happens:
```output
In module 'std_string' imported from /home/johel/Documents/C++/Forks/SFML/SFML/include/SFML/System/String.hpp:35:
/home/johel/root/clang-main/bin/../include/c++/v1/string:723:46: error: implicit instantiation of undefined template 'std::char_traits<unsigned char>'
  723 |     static_assert(( is_same<_CharT, typename traits_type::char_type>::value),
      |                                              ^
/home/johel/Documents/C++/Forks/SFML/SFML/src/SFML/System/String.cpp:171:41: note: in instantiation of template class 'std::basic_string<unsigned char>' requested here
  171 | std::basic_string<std::uint8_t> String::toUtf8() const
      |                                         ^
/home/johel/root/clang-main/bin/../include/c++/v1/__fwd/string.h:23:29: note: template is declared here
   23 | struct _LIBCPP_TEMPLATE_VIS char_traits;
      |                             ^
```

**Comment by ChrisThrasher** [comment]
I actually remember reading that LLVM commit. It would help if you could find the LLVM commit hash that made this change since it may explain why it happened and how we can fix this.

**Comment by JohelEGP** [comment]
Sorry. Bisecting LLVM is something I can't afford to do.

**Comment by JohelEGP** [comment]
I'll see if the history of `<string>` shows anything relevant.
Here's a simpler reproducer: <https://compiler-explorer.com/z/xGW1zbqss>.
```C++
#include <cstdint>
#include <string>
std::basic_string<std::uint8_t> s;
```
```output
In file included from <source>:2:
/opt/compiler-explorer/clang-trunk-20230824/bin/../include/c++/v1/string:723:46: error: implicit instantiation of undefined template 'std::char_traits<unsigned char>'
  723 |     static_assert(( is_same<_CharT, typename traits_type::char_type>::value),
      |                                              ^
<source>:3:33: note: in instantiation of template class 'std::basic_string<unsigned char>' requested here
    3 | std::basic_string<std::uint8_t> s;
      |                                 ^
/opt/compiler-explorer/clang-trunk-20230824/bin/../include/c++/v1/__fwd/string.h:23:29: note: template is declared here
   23 | struct _LIBCPP_TEMPLATE_VIS char_traits;
      |                             ^
1 error generated.
Compiler returned: 1
```
GCC and MSVC accept, so it may be an LLVM issue.

**Comment by ChrisThrasher** [comment]
It's way easier than that. Just search the commit history with a took like `gitk` for references to `std::char_traits<std::uint8_t>` or something similar. I'm sure the commit is from the last few weeks so you're not having to search thousands of commits. Maybe a few hundred tops. We know this change was introduced relatively recently. It would be greatly appreciated and help us fix this bug much faster.

**Comment by texus** [comment]
Probably https://github.com/llvm/llvm-project/commit/e30a148b098d462d0267c479cd9e4783363c2761

**Comment by ChrisThrasher** [comment]
I'm not sure what options we have other than converting all our uses of `std::basic_string<std::uint8_t>` to `std::basic_string<char>`. LLVM 17 isn't even released so we have ~6 months to fix this before LLVM 18 is released. I say that because there's a chance this get removed from LLVM 18 before release. However if LLVM 18 ships with this patch then we're pretty much forced to do something about it.

**Comment by JohelEGP** [comment]
This breaks on Linux:
```output
/home/johel/Documents/C++/Forks/SFML/SFML/src/SFML/Window/Unix/ClipboardImpl.cpp:348:21: error: no matching function for call to 'XChangeProperty'
  348 |                     XChangeProperty(m_display,
      |                     ^~~~~~~~~~~~~~~
/usr/include/X11/Xlib.h:2082:12: note: candidate function not viable: no known conversion from 'const value_type *' (aka 'const char *') to 'const unsigned char *' for 7th argument
 2089 | extern int XChangeProperty(
      |            ^
 2090 |     Display*            /* display */,
 2091 |     Window              /* w */,
 2092 |     Atom                /* property */,
 2093 |     Atom                /* type */,
 2094 |     int                 /* format */,
 2095 |     int                 /* mode */,
 2096 |     _Xconst unsigned char*      /* data */,
      |     ~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
```

**Comment by Quuxplusone** [comment]
Instead of changing `basic_string<uint8_t>` to plain `string` and thus losing the typesystem distinction you worked hard to create in the first place, I would suggest something more like this: https://github.com/Quuxplusone/SFML/commit/alternative-llvm18-string-fix
The idea here is simply to define your own `sf::char_traits_uint8_t` to replace the disappeared `std::char_traits<uint8_t>`, and then use `basic_string<uint8_t, sf::char_traits_uint8_t>` everywhere you previously used `basic_string<uint8_t>`. (I gave it a typedef alias `sf::U8String` as well; that's orthogonal to the main idea of the patch, but IMHO the alias is a good idea no matter whether you like the main idea or not.)

IMO the _best_ best outcome would be if you can persuade @philnik777 to revert the libc++18 change that caused this churn... but I don't claim that that direction is likely to succeed. The question is just how many C++ projects libc++18 is willing to break to make this omelet; I think the answer is probably "lots more than just SFML."

**Comment by philnik777** [comment]
I don't think it's reasonable to revert the change, since it removes the option to easily write really weird code and most likely buggy code (e.g. `std::basic_string<std::array<int, 4>>`). That being said, the main complaint I see is that it breaks `std::basic_string<unsigned char>`. I'm not convinced that this usage makes a lot of sense, since it's pretty much identical to `std::string`, but I also don't see much harm in allowing it. So I think adding a `std::char_traits<unsigned char>` specialization as an extension might be reasonable. Anyways, this isn't the right thread for this. I'll make a patch in a few weeks and talk with Louis about it.

**Comment by JohelEGP** [comment]
FYI: <https://github.com/llvm/llvm-project/pull/66153>.

**Comment by ChrisThrasher** [comment]
> FYI: [llvm/llvm-project#66153](https://github.com/llvm/llvm-project/pull/66153).

Have you confirmed this fixes the compilation errors?

**Comment by JohelEGP** [comment]
No.

**Comment by ldionne** [comment]
FWIW even if we temporarily revert the change, that specialization is deprecated and you folks should find a way to move off of it. `std::basic_string<uint8_t>` is not strictly eStandard C++, that's the bottom line.

**Comment by JohelEGP** [comment]
Now it warns:
```output
[43/252 4 36.926] Building CXX object _deps/sfml-build/src/SFML/System/CMakeFiles/sfml-system.dir/String.cpp.o
In file included from /home/johel/Documents/C++/Forks/SFML/SFML/src/SFML/System/String.cpp:28:
In file included from /home/johel/Documents/C++/Forks/SFML/SFML/include/SFML/System/String.hpp:32:
In file included from /home/johel/Documents/C++/Forks/SFML/SFML/include/SFML/System/Utf.hpp:32:
In file included from /home/johel/Documents/C++/Forks/llvm/libcxx-std-modules/include/c++/v1/locale:200:
In file included from /home/johel/Documents/C++/Forks/llvm/libcxx-std-modules/include/c++/v1/__iterator/istreambuf_iterator.h:14:
In file included from /home/johel/Documents/C++/Forks/llvm/libcxx-std-modules/include/c++/v1/__fwd/istream.h:13:
/home/johel/Documents/C++/Forks/llvm/libcxx-std-modules/include/c++/v1/__fwd/string.h:45:41: warning: 'char_traits<unsigned char>' is deprecated: char_traits<T> for T not equal to char, wchar_t, char8_t, char16_t or char32_t is non-standard and is provided for a temporary period. It will be removed in LLVM 19, so please migrate off of it. [-Wdeprecated-declarations]
   45 | template <class _CharT, class _Traits = char_traits<_CharT>, class _Allocator = allocator<_CharT> >
      |                                         ^
/home/johel/Documents/C++/Forks/SFML/SFML/src/SFML/System/String.cpp:175:10: note: in instantiation of default argument for 'basic_string<std::uint8_t>' required here
  175 |     std::basic_string<std::uint8_t> output;
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~
/home/johel/Documents/C++/Forks/llvm/libcxx-std-modules/include/c++/v1/__string/char_traits.h:81:8: note: 'char_traits<unsigned char>' has been explicitly marked deprecated here
   81 | struct _LIBCPP_DEPRECATED_("char_traits<T> for T not equal to char, wchar_t, char8_t, char16_t or char32_t is non-standard and is provided for a temporary period. It will be removed in LLVM 19, so please migrate off of it.")
      |        ^
/home/johel/Documents/C++/Forks/llvm/libcxx-std-modules/include/c++/v1/__config:922:53: note: expanded from macro '_LIBCPP_DEPRECATED_'
  922 | #      define _LIBCPP_DEPRECATED_(m) __attribute__((__deprecated__(m)))
      |                                                     ^
1 warning generated.
```

**Comment by ChrisThrasher** [comment]
https://github.com/SFML/SFML/actions/runs/7563595762/job/20596308963

I'm trying to fix this but am getting some weird linker errors on Windows. Can anyone help me figure out what's going? It appears that MSVC's standard library defines `char` and `std::uint8_t` to be identical and thus I end up with an ODR violation when trying to specialization `std::char_traits` for `std::uint8_t`. That's my best guess at least.

**Comment by JohelEGP** [comment]
That's not the case: <https://cpp1.godbolt.org/z/fn51q8aPG>.
Seems like a real ODR issue.

**Comment by Quuxplusone** [comment]
> Instead of changing `basic_string<uint8_t>` to plain `string` and thus losing the typesystem distinction you worked hard to create in the first place, I would suggest something more like this: [Quuxplusone@alternative-llvm18-string-fix](https://github.com/Quuxplusone/SFML/commit/alternative-llvm18-string-fix) The idea here is simply to define your own `sf::char_traits_uint8_t` to replace the disappeared `std::char_traits<uint8_t>`, and then use `basic_string<uint8_t, sf::char_traits_uint8_t>` everywhere you previously used `basic_string<uint8_t>`. (I gave it a typedef alias `sf::U8String` as well; that's orthogonal to the main idea of the patch, but IMHO the alias is a good idea no matter whether you like the main idea or not.)

Should I just open a PR for this? I don't know why I didn't do so originally.

**Comment by ChrisThrasher** [comment]
No thanks I've got it covered.

**Comment by ChrisThrasher** [comment]
See https://github.com/SFML/SFML/pull/2885

**Comment by eXpl0it3r** [comment]
What about SFML 2.x? 🤔

**Comment by ChrisThrasher** [comment]
> What about SFML 2.x? 🤔

I'm frankly not sure. Perhaps we can use some preprocessing magic to detect libc++ >=19 and specialize `std::char_traits<std::uint8_t>` in those circumstances?

**Comment by tomi-font** [comment]
I just ran into this trying to compile my project with libc++ 19.
```
/usr/bin/../include/c++/v1/string:820:42: error: implicit instantiation of undefined template 'std::char_traits<unsigned int>'
  820 |   static_assert(is_same<_CharT, typename traits_type::char_type>::value,
      |                                          ^
/usr/include/SFML/System/String.hpp:52:18: note: in instantiation of template class 'std::basic_string<unsigned int>' requested here
   52 |     typedef std::basic_string<Uint32>::iterator       Iterator;      //!< Iterator type
      |                  ^
/usr/bin/../include/c++/v1/__fwd/string.h:23:29: note: template is declared here
   23 | struct _LIBCPP_TEMPLATE_VIS char_traits;
      |                             ^
In file included from /usr/bin/../include/c++/v1/string:647:
/usr/bin/../include/c++/v1/string_view:300:42: error: implicit instantiation of undefined template 'std::char_traits<unsigned int>'
  300 |   static_assert(is_same<_CharT, typename traits_type::char_type>::value,
      |                                          ^
/usr/bin/../include/c++/v1/__iterator/back_insert_iterator.h:78:43: note: in instantiation of template class 'std::basic_string_view<unsigned int>' requested here
   78 |   return back_insert_iterator<_Container>(__x);
      |                                           ^
/usr/include/SFML/System/String.inl:31:36: note: in instantiation of function template specialization 'std::back_inserter<std::basic_string<unsigned int>>' requested here
   31 |     Utf8::toUtf32(begin, end, std::back_inserter(string.m_string));
      |                                    ^
/usr/bin/../include/c++/v1/__fwd/string.h:23:29: note: template is declared here
   23 | struct _LIBCPP_TEMPLATE_VIS char_traits;
      |                             ^
```
Using a pre-compiled SFML 2.6.1, it actually works fine with libc++ 18 but breaks with 19.
It would have been nice to have this fixed for 2.6 as well.
I see there's now the first RC for SFML 3 (congrats!), so it should be possible to jump to it to start using libc++ 19.
